File: NotesDesk/notes-desk-client/win/NotesDeskClient.Win/Services/AppSettings.cs
using System.IO;
using System.Text.Json;
using NotesDeskClient.Win.Models;

namespace NotesDeskClient.Win.Services;

public static class AppSettings
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    private static string SettingsPath =>
        Path.Combine(AppPaths.AppDataDirectory, "settings.json");

    private static AppSettingsData Load()
    {
        try
        {
            if (File.Exists(SettingsPath))
            {
                var json = File.ReadAllText(SettingsPath);
                return JsonSerializer.Deserialize<AppSettingsData>(json) ?? new AppSettingsData();
            }
        }
        catch { /* ignore */ }
        return new AppSettingsData();
    }

    private static void Save(AppSettingsData data)
    {
        Directory.CreateDirectory(AppPaths.AppDataDirectory);
        File.WriteAllText(SettingsPath, JsonSerializer.Serialize(data, JsonOptions));
    }

    public static string ServerBaseUrl
    {
        get => Load().ServerBaseUrl ?? "http://127.0.0.1:8080";
        set { var d = Load(); d.ServerBaseUrl = value.Trim().TrimEnd('/'); Save(d); }
    }

    public static string ApiToken
    {
        get => Load().ApiToken ?? "";
        set { var d = Load(); d.ApiToken = value; Save(d); }
    }

    public static double PanelOpacity
    {
        get => Load().PanelOpacity <= 0 ? 0.92 : Load().PanelOpacity;
        set { var d = Load(); d.PanelOpacity = value; Save(d); }
    }

    public static double FontSize
    {
        get => Load().FontSize <= 0 ? 14 : Load().FontSize;
        set { var d = Load(); d.FontSize = value; Save(d); }
    }

    public static bool PanelVisible
    {
        get => Load().PanelVisible ?? true;
        set { var d = Load(); d.PanelVisible = value; Save(d); }
    }

    public static DeskViewMode ViewMode
    {
        get => Enum.TryParse<DeskViewMode>(Load().ViewMode, out var m) ? m : DeskViewMode.Incomplete;
        set { var d = Load(); d.ViewMode = value.ToString(); Save(d); }
    }

    public static bool ShowCompletedSection
    {
        get => Load().ShowCompletedSection ?? true;
        set { var d = Load(); d.ShowCompletedSection = value; Save(d); }
    }

    public static bool QuitOnClosePanel
    {
        get => Load().QuitOnClosePanel ?? false;
        set { var d = Load(); d.QuitOnClosePanel = value; Save(d); }
    }

    public static (double X, double Y)? PanelOrigin
    {
        get
        {
            var d = Load();
            if (d.PanelOriginX.HasValue && d.PanelOriginY.HasValue)
                return (d.PanelOriginX.Value, d.PanelOriginY.Value);
            return null;
        }
        set
        {
            var d = Load();
            if (value.HasValue)
            {
                d.PanelOriginX = value.Value.X;
                d.PanelOriginY = value.Value.Y;
            }
            else
            {
                d.PanelOriginX = null;
                d.PanelOriginY = null;
            }
            Save(d);
        }
    }

    public static bool IsConfigured => !string.IsNullOrWhiteSpace(ServerBaseUrl);

    private sealed class AppSettingsData
    {
        public string? ServerBaseUrl { get; set; }
        public string? ApiToken { get; set; }
        public double PanelOpacity { get; set; }
        public double FontSize { get; set; }
        public bool? PanelVisible { get; set; }
        public string? ViewMode { get; set; }
        public bool? ShowCompletedSection { get; set; }
        public bool? QuitOnClosePanel { get; set; }
        public double? PanelOriginX { get; set; }
        public double? PanelOriginY { get; set; }
    }
}

public static class AppPaths
{
    public static string AppDataDirectory =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            AppConstants.AppSupportSubdirectory);
}

File: NotesDesk/notes-desk-client/mac/Sources/NotesDeskClient/Utilities/AppSettings.swift
import Foundation

enum AppSettings {
    private enum Key {
        static let panelOpacity = "panelOpacity"
        static let fontSize = "fontSize"
        static let panelOriginX = "panelOriginX"
        static let panelOriginY = "panelOriginY"
        static let panelVisible = "panelVisible"
        static let viewMode = "viewMode"
        static let showCompletedSection = "showCompletedSection"
        static let quitOnClosePanel = "quitOnClosePanel"
        static let serverBaseURL = "serverBaseURL"
        static let apiToken = "apiToken"
        static let sessionCookie = "sessionCookie"
        static let webUsername = "webUsername"
        static let webPassword = "webPassword"
    }

    static var panelOpacity: Double {
        get {
            let value = UserDefaults.standard.double(forKey: Key.panelOpacity)
            return value == 0 ? 0.92 : value
        }
        set { UserDefaults.standard.set(newValue, forKey: Key.panelOpacity) }
    }

    static var fontSize: Double {
        get {
            let value = UserDefaults.standard.double(forKey: Key.fontSize)
            return value == 0 ? 14 : value
        }
        set { UserDefaults.standard.set(newValue, forKey: Key.fontSize) }
    }

    static var panelOrigin: CGPoint {
        get {
            CGPoint(
                x: UserDefaults.standard.double(forKey: Key.panelOriginX),
                y: UserDefaults.standard.double(forKey: Key.panelOriginY)
            )
        }
        set {
            UserDefaults.standard.set(newValue.x, forKey: Key.panelOriginX)
            UserDefaults.standard.set(newValue.y, forKey: Key.panelOriginY)
        }
    }

    static var panelVisible: Bool {
        get {
            if UserDefaults.standard.object(forKey: Key.panelVisible) == nil {
                return true
            }
            return UserDefaults.standard.bool(forKey: Key.panelVisible)
        }
        set { UserDefaults.standard.set(newValue, forKey: Key.panelVisible) }
    }

    static var viewMode: DeskViewMode {
        get {
            guard let raw = UserDefaults.standard.string(forKey: Key.viewMode),
                  let mode = DeskViewMode(rawValue: raw)
            else {
                return .incomplete
            }
            return mode
        }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: Key.viewMode) }
    }

    static var showCompletedSection: Bool {
        get {
            if UserDefaults.standard.object(forKey: Key.showCompletedSection) == nil {
                return true
            }
            return UserDefaults.standard.bool(forKey: Key.showCompletedSection)
        }
        set { UserDefaults.standard.set(newValue, forKey: Key.showCompletedSection) }
    }

    static var quitOnClosePanel: Bool {
        get { UserDefaults.standard.bool(forKey: Key.quitOnClosePanel) }
        set { UserDefaults.standard.set(newValue, forKey: Key.quitOnClosePanel) }
    }

    /// 如 http://192.168.1.100:8080
    static var serverBaseURL: String {
        get { UserDefaults.standard.string(forKey: Key.serverBaseURL) ?? "http://127.0.0.1:8080" }
        set { UserDefaults.standard.set(newValue.trimmingCharacters(in: .whitespacesAndNewlines), forKey: Key.serverBaseURL) }
    }

    static var apiToken: String {
        get { UserDefaults.standard.string(forKey: Key.apiToken) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Key.apiToken) }
    }

    /// Web 登录 Session（/api/auth/login 后写入）
    static var sessionCookie: String {
        get { UserDefaults.standard.string(forKey: Key.sessionCookie) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Key.sessionCookie) }
    }

    static var webUsername: String {
        get { UserDefaults.standard.string(forKey: Key.webUsername) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Key.webUsername) }
    }

    static var webPassword: String {
        get { UserDefaults.standard.string(forKey: Key.webPassword) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Key.webPassword) }
    }

    static var hasApiAuth: Bool {
        !apiToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || !sessionCookie.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static var isConfigured: Bool {
        !serverBaseURL.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}
